File: python/src/agent_framework/hitl.py
"""Human-in-the-loop modes and types."""

from enum import Enum
from typing import Literal


class HumanInTheLoopMode(str, Enum):
    """HITL modes: ALWAYS, TERMINATE, NEVER."""

    ALWAYS = "ALWAYS"
    TERMINATE = "TERMINATE"
    NEVER = "NEVER"


InterceptDecision = Literal["approve", "skip", "terminate", "modify"]


class InterceptContext:
    """Context passed to intercept handler."""

    def __init__(
        self,
        thread_id: str,
        node_id: str,
        step: int,
        state: dict,
        proposed_action: str | None = None,
        proposed_output: object = None,
        timestamp: float = 0,
    ):
        self.thread_id = thread_id
        self.node_id = node_id
        self.step = step
        self.state = state
        self.proposed_action = proposed_action
        self.proposed_output = proposed_output
        self.timestamp = timestamp


class InterceptResponse:
    """User response from intercept."""

    def __init__(
        self,
        decision: InterceptDecision,
        modified_state: dict | None = None,
        reason: str | None = None,
    ):
        self.decision = decision
        self.modified_state = modified_state
        self.reason = reason

File: packages/core/src/hitl.ts
import { z } from "zod";

/**
 * Human-in-the-loop modes (AutoGen-style):
 * - ALWAYS: Intercept before every step; user can approve/skip/terminate
 * - TERMINATE: Intercept only at end conditions (e.g. before final output)
 * - NEVER: Fully autonomous; no interception
 */
export const HumanInTheLoopModeSchema = z.enum(["ALWAYS", "TERMINATE", "NEVER"]);
export type HumanInTheLoopMode = z.infer<typeof HumanInTheLoopModeSchema>;

/**
 * User decision at an intercept point.
 */
export const InterceptDecisionSchema = z.enum(["approve", "skip", "terminate", "modify"]);
export type InterceptDecision = z.infer<typeof InterceptDecisionSchema>;

/**
 * Context passed to the intercept handler for user approval.
 */
export interface InterceptContext {
  threadId: string;
  nodeId: string;
  step: number;
  state: Record<string, unknown>;
  proposedAction?: string;
  proposedOutput?: unknown;
  timestamp: number;
}

/**
 * User response from an intercept (approve/skip/terminate/modify).
 */
export interface InterceptResponse {
  decision: InterceptDecision;
  /** Optional modified state when decision is "modify" */
  modifiedState?: Record<string, unknown>;
  /** Optional reason for audit */
  reason?: string;
}
